test task by @Comediant24

Comediant24 opened the pull request:
## Добрый день!

#### Кратко о фичах:

- Для отрисовки карточек создан template элемент - так быстрее, такой код легче читать, легче поддерживать.
- На попапы повешены обработчики закрытия попапа: по клавише, по клику на оверлей.
- После закрытия попапа обработчик клавиши Esc удаляется.
- Файловая структура и верстка выполнена по БЭМ.
- Использованы семантические теги.
- Для Popup и Card сделаны классы и для их отрисовки используются соответствующие методы классов.
- Не использую модули так как по ТЗ нужно запускать без сервера.

Comediant24 committed:
feat: создать базовую разметку и разметку карточки

Добавить normalise.css

Добавить стили и разметку для тела страницы

Добавить template разметку и стили для карточки товара

Использовать для файловой структуры БЭМ

Comediant24 committed:
feat: добавить класс Card

Добавить класс Card для рендера карт товаров

Добавить метод классу Card для генерации карточки

Реализовать отрисовку карточек на страницу

Comediant24 committed:
feat: добавить класс Popup и PopupBasket

Создать стили для Popup и PopupBasket

Создать логику работы PopupBasket

Добавить функцию открытия попапа корзины

Добавить обработчик кнопки корзины на карточку

Добавить в constants конфиги Popup

Comediant24 committed:
feat: добавить класс PopupOrder

Создать стили для PopupOrder

Создать логику работы PopupOrder

Добавить функцию открытия попапа заказа

Добавить обработчик кнопки заказа на карточку

Обновит в constants конфиги Popup

Добавить атрибуты type и aria-label на кнопки

Comediant24 committed:
feat: отрефакторить код

Добавить title html

Перенести обрабочики открытия Popup в колбек new Card

Удалить дублирующийся код в PopupBasket и PopupOrder,
добавить обработчик на кнопку в корзину

Переименовать константы в конфиге POPUP_CONFIG

Comediant24 committed:
feat: улучшить UX popup корзины

Удалить overlay попапа корзины

Добавить фичу закрытие попапа корзины через 5 секунд

Перенести логику закрытия попапа по overlay в PopupOrder

idbolshakov commented on src/index.js:
```diff
@@ -0,0 +1,17 @@
+
+API.products.forEach((card) => {
+  const cardProduct = new Card(TEMPLATE_CARD, card, {
+    handleBasketClick: () => {
```
для каждой карточки свои обработчики событий - неэффективно

Comediant24 replied:
да, решил это через делегирование.

idbolshakov commented on src/index.js:
```diff
@@ -0,0 +1,17 @@
+    },
+  });
+
+  list.append(cardProduct.generateCard());
```
10 раз будет вызываться дорогая операция вставки элементов в DOM

idbolshakov commented on src/utils/constants.js:
```diff
@@ -0,0 +1,17 @@
+var TEMPLATE_CARD = 'card-template';
```
объявление кучи переменных в глобальной области видимости - плохая практика

idbolshakov reviewed (changes requested):
Привет.  Спасибо за выполнение ТЗ. В общем и целом все круто. Видно что знаешь что такое БЭМ и для чего он, а также приятна общая тенденция к структуризации кода, при этом код не выглядит перегруженным.

Есть небольшие замечания:
- очень перегруженный файл со стилями блока popup, кажется что лучше разбить (сейчас этого делать не нужно)
- история с использованием `<template>` тега лично для меня выглядит не очень красиво, в будущем такой код будет сложно поддерживать. Редактирование шаблона выполнено в очень императивном стиле. Если нужно добавить какой-то динамически изменяемый атрибут в теге, тебе нужно сначала добавить его в html, потом внутри js кода написать строчку с редактированием этого атрибута.
- пустой класс `Product` зачем?
- по производительности вариант с навшиванием слушателей событий под каждый экземпляр класса `Card` очень плохой. С одной стороны ты хотел инкапсулировать всю логику работы  с карточкой товара внутрь класса, но фактически это решение приводит к просадкам производительности.  
- повторяющиеся цвета можно вынести в css переменные
- в файле popup.js три класса объявлены - не очень красиво

Comediant24 committed:
feat: создать обработчик событий на списке

Убрать слушатель событий с кажого экзембляра класса Card

Создать слушатель на всем списке карточек и в нем калбек с делегированием
событий

Создать функцию getCardData для получения данных карточки с целевого
события

Comediant24 committed:
fix: удалить блокирующий overlay на popupBasket

Удалить значения max и min ширины для оверлея popup_basket
Оверлей наклыдвался слоём и блокировал доступ к интерфейсу

Удалить пустой класс Product

Comediant24 commented:
Вчера уже после подумал, что естественно я бы этот обработчик реализовал бы через делегирование.
В новых коммитах убрал обработчики с каждой карточки и пофиксил баг с оверлеем.
